File: solutions/haskell/protein-translation/1/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where
import Data.Map (Map)
import qualified Data.Map as M

proteinMap :: Map String String
proteinMap = M.fromList [ ("AUG","Methionine")
                        , ("UUU","Phenylalanine")
                        , ("UUC","Phenylalanine")
                        , ("UUA","Leucine")
                        , ("UUG","Leucine")
                        , ("UCU","Serine") 
                        , ("UCC","Serine")
                        , ("UCA","Serine")
                        , ("UCG","Serine")
                        , ("UAU","Tyrosine")
                        , ("UAC","Tyrosine")
                        , ("UGU","Cysteine")
                        , ("UGC","Cysteine")
                        , ("UGG","Tryptophan")]

proteins :: String -> Maybe [String]
proteins = sequenceA . proteins'

proteins' :: String -> [Maybe String]
proteins' (x:y:z:xs) = maybe [] (\x -> Just x : proteins' xs) $ M.lookup [x,y,z] proteinMap
proteins' _ = []

File: solutions/fortran/protein-translation/1/protein_translation.f90
module protein_translation
  implicit none

contains

  function proteins(rna) result(names)
    character(len=*), intent(in) :: rna
    character(len=3) :: codon
    character(len=13), allocatable :: names(:)
    integer :: names_length
    integer :: i

    names_length = 0
    do i = 1, len(rna), 3
      if (rna(i:i+2).eq."UAA".or.rna(i:i+2).eq."UAG".or.rna(i:i+2).eq."UGA") then
        exit
      end if
      names_length = names_length + 1
    end do
    allocate(names(names_length))
    do i = 1, names_length
      codon = rna((((i-1)*3)+1):(((i-1)*3)+3))
      if (codon.eq."AUG") then
        names(i) = "Methionine"
      else if (codon.eq."UUU".or.codon.eq."UUC") then
        names(i) = "Phenylalanine"
      else if (codon.eq."UUA".or.codon.eq."UUG") then
        names(i) = "Leucine"
      else if (codon.eq."UCU".or.codon.eq."UCC".or.codon.eq."UCA".or.codon.eq."UCG") then
        names(i) = "Serine"
      else if (codon.eq."UAU".or.codon.eq."UAC") then
        names(i) = "Tyrosine"
      else if (codon.eq."UGU".or.codon.eq."UGC") then
        names(i) = "Cysteine"
      else if (codon.eq."UGG") then
        names(i) = "Tryptophan"
      else
        exit
      end if
    end do
  end function proteins

end module protein_translation

